### DISCUSSION TITLE
Inactive kernel gets executed

### DISCUSSION THREAD

**Discussion by Richardk2n** [pyopencl/pyopencl#798] [Q&A] [answered]
This is more of a general OpenCL question rather than a pyopencl question, but you seem to know a lot, so I hope asking here is ok.

Why does this not work:
```python
import pyopencl as cl

plattform = cl.get_platforms()[0]
device = plattform.get_devices()[0]

ctx = cl.Context([device])
queue = cl.CommandQueue(ctx)

src = r"""
kernel void tet() {
    const long m = get_global_id(0);
	printf("ID %ld\n", m);
}
"""

src2 = r"""
kernel void other() {
    const long n = get_global_id(0);
    printf("Do not print %ld\n", n);
}
"""

p = cl.Program(ctx, src).compile()
p2 = cl.Program(ctx, src2).compile()

program = cl.link_program(ctx, [p, p2])

program.tet(queue, (2,), None)
```

I would expect this to output:
```
ID 0
ID 1
```

But what I get is:
```
ID 0
Do not print 4294967297
(null)
```

This seems to be specifically related to the `printf` statements. Commenting out the one in `tet` removes all output, commenting the one in `other` results in output as expected. Is what I am doing forbidden in general or should this work?

**Comment by inducer** [comment]
I can't think of a reason why `other` would be called, or why the `printf` in it would execute. In fact, executing your reproducer with pocl 6.0 gives what you list as the expected output.

**Reply by Richardk2n** [reply]
Thank you for your input. So I installed pocl and ran this code on my CPU, and it worked as expected. So does that mean, that the issue is in the opencl lib from amd/nvidia, and I have to ask them?

**Reply by inducer** [reply, marked as answer]
That's my read of the situation, yes. Note that pocl can also take to Nvidia GPUs through its CUDA support, though I am uncertain whether `printf` is supported there.
